File: android/core/src/main/kotlin/com/allthingsclaude/battery/core/BurnRateCalculator.kt
package com.allthingsclaude.battery.core

import java.time.Instant
import kotlin.math.abs

/**
 * A point-in-time usage sample. On macOS these are persisted to SQLite; on the
 * phone a small ring buffer of them exists purely to feed the regression, so no
 * database is needed. Field set is the subset [BurnRateCalculator] reads.
 *
 * Port of `ios/BatteryKit/UsageSnapshot.swift`.
 */
data class UsageSnapshot(
    val timestamp: Instant,
    val sessionUtilization: Double,
    val sessionResetsAt: Instant,
)

/** A burn-rate projection derived from recent snapshots. */
data class BurnRateProjection(
    /** Percentage points per hour. */
    val currentRate: Double,
    /** When 100% is hit; null if flat or already maxed. */
    val projectedLimitTime: Instant?,
    /** Projected utilization at reset time. */
    val projectedAtReset: Double,
    val trend: Trend,
    val hasEnoughData: Boolean,
) {
    enum class Trend { INCREASING, STABLE, DECREASING }

    companion object {
        fun insufficientData(utilization: Double) = BurnRateProjection(
            currentRate = 0.0,
            projectedLimitTime = null,
            projectedAtReset = utilization,
            trend = Trend.STABLE,
            hasEnoughData = false,
        )
    }
}

/**
 * Linear-regression burn rate. Ported from `ios/BatteryKit/BurnRateCalculator.swift`,
 * which is itself a port of the macOS original — the arithmetic is identical on
 * all three platforms and `core/src/test/resources/fixtures/` exists to keep it
 * that way.
 */
object BurnRateCalculator {

    const val MINIMUM_SNAPSHOTS = 3

    /** Seconds. Below this the samples are too bunched for a slope to mean anything. */
    const val MINIMUM_TIME_SPAN = 120.0

    /**
     * Tolerance when matching a snapshot's window to the current one. A reset
     * time round-trips through JSON and can come back a hair off, so an exact
     * comparison would silently discard the entire history.
     */
    private const val RESET_MATCH_TOLERANCE_SECONDS = 1.0

    fun calculate(
        snapshots: List<UsageSnapshot>,
        currentUtilization: Double,
        resetsAt: Instant,
        now: Instant = Instant.now(),
    ): BurnRateProjection {
        // Only samples belonging to the current session window: a reset changes
        // `sessionResetsAt`, and mixing windows would regress across the
        // discontinuity and report a wildly negative or positive pace.
        val currentSession = snapshots.filter {
            abs(secondsBetween(it.sessionResetsAt, resetsAt)) < RESET_MATCH_TOLERANCE_SECONDS
        }

        if (currentSession.size < MINIMUM_SNAPSHOTS) {
            return BurnRateProjection.insufficientData(currentUtilization)
        }

        val sorted = currentSession.sortedBy { it.timestamp }
        val timeSpan = secondsBetween(sorted.last().timestamp, sorted.first().timestamp)
        if (timeSpan < MINIMUM_TIME_SPAN) {
            return BurnRateProjection.insufficientData(currentUtilization)
        }

        val origin = sorted.first().timestamp.toEpochMilli() / 1000.0
        val n = sorted.size.toDouble()
        var sumX = 0.0; var sumY = 0.0; var sumXY = 0.0; var sumX2 = 0.0
        for (s in sorted) {
            val x = (s.timestamp.toEpochMilli() / 1000.0 - origin) / 3600.0 // hours
            val y = s.sessionUtilization
            sumX += x; sumY += y; sumXY += x * y; sumX2 += x * x
        }

        val denominator = n * sumX2 - sumX * sumX
        if (abs(denominator) <= 1e-10) {
            return BurnRateProjection.insufficientData(currentUtilization)
        }

        // Within a session, utilization only rises — clamp regression noise (and
        // a genuine post-reset drop that slipped the window filter) to zero
        // rather than reporting a negative pace nothing downstream can use.
        val slope = maxOf(0.0, (n * sumXY - sumX * sumY) / denominator)
        val trend = if (slope > 1.0) BurnRateProjection.Trend.INCREASING else BurnRateProjection.Trend.STABLE

        var projectedLimitTime: Instant? = null
        if (slope > 0.01 && currentUtilization < 100) {
            val hoursToLimit = (100.0 - currentUtilization) / slope
            projectedLimitTime = now.plusMillis((hoursToLimit * 3600 * 1000).toLong())
        }

        val hoursToReset = maxOf(0.0, secondsBetween(resetsAt, now) / 3600.0)
        val projectedAtReset = (currentUtilization + slope * hoursToReset).coerceIn(0.0, 100.0)

        return BurnRateProjection(
            currentRate = slope,
            projectedLimitTime = projectedLimitTime,
            projectedAtReset = projectedAtReset,
            trend = trend,
            hasEnoughData = true,
        )
    }

    private fun secondsBetween(a: Instant, b: Instant): Double =
        (a.toEpochMilli() - b.toEpochMilli()) / 1000.0
}

File: ios/BatteryKit/BurnRateCalculator.swift
import Foundation

/// A burn-rate projection derived from recent snapshots. Mirrors the macOS type
/// so any code that reasons about "will I hit the limit" behaves identically.
struct BurnRateProjection {
    let currentRate: Double           // percentage points per hour
    let projectedLimitTime: Date?     // when 100% is hit (nil if flat / already maxed)
    let projectedAtReset: Double      // projected utilization at reset time
    let trend: Trend
    let hasEnoughData: Bool

    enum Trend: String { case increasing, stable, decreasing }

    static func insufficientData(at utilization: Double) -> BurnRateProjection {
        BurnRateProjection(
            currentRate: 0,
            projectedLimitTime: nil,
            projectedAtReset: utilization,
            trend: .stable,
            hasEnoughData: false
        )
    }
}

/// Linear-regression burn rate. Ported from the macOS `BurnRateCalculator`; the
/// only difference is the iOS snapshot has fewer fields, but the math is the same.
enum BurnRateCalculator {

    static let minimumSnapshots = 3
    static let minimumTimeSpan: TimeInterval = 120 // 2 minutes

    static func calculate(
        snapshots: [UsageSnapshot],
        currentUtilization: Double,
        resetsAt: Date
    ) -> BurnRateProjection {
        // Only keep samples belonging to the current session window; a reset
        // changes `sessionResetsAt` and would otherwise poison the regression.
        let currentSession = snapshots.filter {
            abs($0.sessionResetsAt.timeIntervalSince(resetsAt)) < 1.0
        }

        guard currentSession.count >= minimumSnapshots else {
            return .insufficientData(at: currentUtilization)
        }

        let sorted = currentSession.sorted { $0.timestamp < $1.timestamp }
        let timeSpan = sorted.last!.timestamp.timeIntervalSince(sorted.first!.timestamp)
        guard timeSpan >= minimumTimeSpan else {
            return .insufficientData(at: currentUtilization)
        }

        let origin = sorted.first!.timestamp.timeIntervalSince1970
        let n = Double(sorted.count)
        var sumX = 0.0, sumY = 0.0, sumXY = 0.0, sumX2 = 0.0
        for s in sorted {
            let x = (s.timestamp.timeIntervalSince1970 - origin) / 3600.0 // hours
            let y = s.sessionUtilization
            sumX += x; sumY += y; sumXY += x * y; sumX2 += x * x
        }

        let denominator = n * sumX2 - sumX * sumX
        guard abs(denominator) > 1e-10 else {
            return .insufficientData(at: currentUtilization)
        }

        // Within a session, utilization only rises — clamp noise to non-negative.
        let slope = max(0, (n * sumXY - sumX * sumY) / denominator)
        let trend: BurnRateProjection.Trend = slope > 1.0 ? .increasing : .stable

        var projectedLimitTime: Date?
        if slope > 0.01 && currentUtilization < 100 {
            let hoursToLimit = (100.0 - currentUtilization) / slope
            projectedLimitTime = Date().addingTimeInterval(hoursToLimit * 3600)
        }

        let hoursToReset = max(0, resetsAt.timeIntervalSinceNow / 3600.0)
        let projectedAtReset = min(100, max(0, currentUtilization + slope * hoursToReset))

        return BurnRateProjection(
            currentRate: slope,
            projectedLimitTime: projectedLimitTime,
            projectedAtReset: projectedAtReset,
            trend: trend,
            hasEnoughData: true
        )
    }
}
